thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java

```diff
@@ -1,274 +1,268 @@
-    detectionJobContext.setAnomalyFunctionId(anomalyFunctionSpec.getId());
-    detectionJobContext.setJobName(jobKey);
+  public Long runAdhocAnomalyFunction(Long functionId, Long startTime, Long endTime) {
+    Long jobExecutionId = null;
 
-    job.getJobDataMap().put(DetectionJobRunner.DETECTION_JOB_CONTEXT, detectionJobContext);
-    job.getJobDataMap().put(DetectionJobRunner.DETECTION_JOB_MONITORING_WINDOW_START_TIME, windowStartTime);
-    job.getJobDataMap().put(DetectionJobRunner.DETECTION_JOB_MONITORING_WINDOW_END_TIME, windowEndTime);
+    AnomalyFunctionDTO anomalyFunction = DAO_REGISTRY.getAnomalyFunctionDAO().findById(functionId);
 
+    String dataset = anomalyFunction.getCollection();
+    DatasetConfigDTO datasetConfig = null;
     try {
-      quartzScheduler.scheduleJob(job, trigger);
-      LOG.info("Started {}: {}", jobKey, anomalyFunctionSpec);
-    } catch (SchedulerException e) {
-      LOG.error("Exception while scheduling job", e);
+      datasetConfig = CACHE_REGISTRY.getDatasetConfigCache().get(dataset);
+    } catch (ExecutionException e) {
+      LOG.error("Exception in fetching dataset config", e);
     }
 
-    return jobKey;
-  }
-
-  private void scheduleJob(JobContext jobContext, AnomalyFunctionDTO anomalyFunctionSpec) {
-
-    String triggerKey = String.format("anomaly_scheduler_trigger_%d", anomalyFunctionSpec.getId());
-    CronTrigger trigger =
-        TriggerBuilder.newTrigger().withIdentity(triggerKey)
-            .withSchedule(CronScheduleBuilder.cronSchedule(anomalyFunctionSpec.getCron())
-                .inTimeZone(TimeZone.getTimeZone("UTC"))).build();
-
-    String jobKey = jobContext.getJobName();
-    JobDetail job = JobBuilder.newJob(DetectionJobRunner.class).withIdentity(jobKey).build();
-
-    job.getJobDataMap().put(DetectionJobRunner.DETECTION_JOB_CONTEXT, jobContext);
-
-    try {
-      quartzScheduler.scheduleJob(job, trigger);
-    } catch (SchedulerException e) {
-      LOG.error("Exception while scheduling job", e);
+    boolean pass = checkIfDetectionRunCriteriaMet(startTime, endTime, datasetConfig, anomalyFunction);
+    if (pass) {
+      jobExecutionId = runAnomalyFunctionOnRanges(anomalyFunction, Lists.newArrayList(startTime), Lists.newArrayList(endTime));
+    } else {
+      LOG.warn("Data incomplete for monitoring window {} ({}) to {} ({}), skipping anomaly detection",
+          startTime, new DateTime(startTime), endTime, new DateTime(endTime));
+      // TODO: Send email to owners/dev team
     }
-
-    LOG.info("Started {}: {}", jobKey, anomalyFunctionSpec);
+    return jobExecutionId;
   }
 
-  private List<AnomalyFunctionDTO> readAnomalyFunctionSpecs() {
-    return DAO_REGISTRY.getAnomalyFunctionDAO().findAll();
+  /**
+   * Checks if a time range for a dataset meets data completeness criteria
+   * @param startTime
+   * @param endTime
+   * @param datasetConfig
+   * @param anomalyFunction
+   * @return
```

Comment:
"return true if the criteria is met or the dataset and anomaly function do not request for data quality check"